packages/loader/container-loader/src/container.ts

```diff
@@ -816,26 +816,45 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
             return;
         }
 
-        const previousContextState = await this.context.snapshotRuntimeState();
+        // By default, close the container and let the host reload.
+        // If hotSwapContext is true in the loader options, then try
+        // to reload the context without closing the container.
+        type ReloadState = { hotSwap: true; prevState: IRuntimeState } | { hotSwap: false };
+        let state: ReloadState = { hotSwap: false };
+        const loaderOptionsHotSwap = this.options.hotSwapContext as boolean ?? false;
+        if (loaderOptionsHotSwap) {
+            const prevState = await this.context.snapshotRuntimeState();
+            state = { hotSwap: true, prevState };
+        }
         this.context.dispose(new Error("ContextDisposedForReload"));
 
-        // don't fire this event if we are transitioning from a null runtime to a real runtime
+        // We always hot-swap, but we don't fire the contextDisposed event
+        // if we are transitioning from a null runtime to a real runtime
         // with detached container we no longer need the null runtime, but for legacy
         // reasons need to keep it around (old documents without summary before code proposal).
         // client's shouldn't need to care about this transition, as it is a implementation detail.
         // if we didn't do this check, the clients would need to do it themselves,
         // which would futher spread the usage of the hasNullRuntime property
         // making it harder to deprecate.
-        if (!this.hasNullRuntime()) {
+        if (this.hasNullRuntime()) {
+            if (!state.hotSwap) {
+                state = { hotSwap: true, prevState: {} };
+            }
+        } else {
             this.emit("contextDisposed", codeDetails, this.context?.codeDetails);
         }
+
         if (this.closed) {
             return;
         }
+        if (!state.hotSwap) {
```

Comment:
for back compat we need to support hot swapping the null runtime

Reply:
This is taken care of a few lines below. We check if `this.hasNullRuntime()` and then force `state.hotSwap` to true if so.